module/zfs/arc.c

```diff
@@ -8531,7 +8532,8 @@ l2arc_apply_transforms(spa_t *spa, arc_buf_hdr_t *hdr, uint64_t asize,
 		cabd = abd_alloc_for_io(asize, ismd);
 		tmp = abd_borrow_buf(cabd, asize);
 
-		psize = zio_compress_data(compress, to_write, tmp, size);
+		psize = zio_compress_data(compress, to_write, tmp, size,
+		    1 << spa->spa_max_ashift);
```

Comment:
This case seems similar to arc_hdr_authenticate(), in that we know the data was previously compressed, and we are trying to recreate the same on-disk bits that we had before.  So I think we should pass `compress_threshold=0` here.